### ISSUE TITLE
Incorrect postconditions for non-result ensures

### ISSUE THREAD

**Issue by SergeyTeplyakov** [Microsoft/CodeContracts#111]
Consider following code:

```c#
private static bool ShouldBeTrue;
public async static Task<string> FooAsync()
{
    Contract.Ensures(Contract.Result<string>() != null);
    Contract.Ensures(ShouldBeTrue);

    ShouldBeTrue = false;

    await Task.Delay(42);

    ShouldBeTrue = true;

    return null;
}
```

For this code ccrewrite will generate following:

```c#
[CompilerGenerated]
		private class <FooAsync>AsyncContractClosure_0
		{
			public string CheckPost(Task<string> task)
			{
				string result = task.Result;
				if (__ContractsRuntime.insideContractEvaluation <= 4)
				{
					try
					{
						__ContractsRuntime.insideContractEvaluation++;
						__ContractsRuntime.Ensures(task.Result != null, null, "Contract.Result<string>() != null");
					}
					finally
					{
						__ContractsRuntime.insideContractEvaluation--;
					}
				}
				return result;
			}
		}
		private static bool ShouldBeTrue;
		public static Task<string> FooAsync()
		{
			Program.<FooAsync>AsyncContractClosure_0 @object = new Program.<FooAsync>AsyncContractClosure_0();
			Program.ShouldBeTrue = false;
			Program.ShouldBeTrue = true;
			Task<string> task = Task.FromResult<string>(null);
			Task<string> task2 = task;
			__ContractsRuntime.Ensures(Program.ShouldBeTrue, null, "ShouldBeTrue");
			Func<Task<string>, string> continuationFunction = new Func<Task<string>, string>(@object.CheckPost);
			return task2.ContinueWith<string>(continuationFunction);
		}
```

I.e. `Contract.Ensures(Contract.Result<string>() != null);` is checked asynchronously via home-backed closure (in the method `@object.CheckPost`), but `Contract.Ensures(ShouldBeTrue);` is checked synchronously.

This code is not correct and both postconditions should be checked in the task continuation.

**Comment by sharwell** [comment]
> This code is not correct and both postconditions should be checked in the task continuation.

First of all, unless `Contract.Ensures` calls are ordered with respect to each other, this code is not *incorrect* -- at least not for the example provided above. In the example, both `Ensures` checks occur after the last line of code in `FooAsync` but before the result of `FooAsync` can be observed by a caller. This is correct behavior unless ordering guarantees or something else places further restrictions on the way postconditions are verified.

Either way, why not check both synchronously? By this I mean either avoid a continuation where possible, or use `TaskContinuationOptions.ExecuteSynchronously`.

**Comment by SergeyTeplyakov** [comment]
@sharwell Ok, let me clarify this issue.

First of all specified method fails with postcondition violation. Obviously this method should not do this. Right :question: 

Second, I've posted incorrect decompiled version of the code. Here is an updated version of the `FooAsync` implementation (this is true for all compilers VS2012-VS2015):

```c#
		private static bool ShouldBeTrue;
		[DebuggerStepThrough, AsyncStateMachine(typeof(Program.<FooAsync>d__0))]
		public static Task<string> FooAsync()
		{
			Program.<FooAsync>AsyncContractClosure_0 @object = new Program.<FooAsync>AsyncContractClosure_0();
			Program.<FooAsync>d__0 <FooAsync>d__;
			<FooAsync>d__.<>t__builder = AsyncTaskMethodBuilder<string>.Create();
			<FooAsync>d__.<>1__state = -1;
			AsyncTaskMethodBuilder<string> <>t__builder = <FooAsync>d__.<>t__builder;
			<>t__builder.Start<Program.<FooAsync>d__0>(ref <FooAsync>d__);
			Task<string> task = <FooAsync>d__.<>t__builder.Task;
			Task<string> task2 = task;
			__ContractsRuntime.Ensures(Program.ShouldBeTrue, null, "ShouldBeTrue");
			Func<Task<string>, string> continuationFunction = new Func<Task<string>, string>(@object.CheckPost);
			return task2.ContinueWith<string>(continuationFunction);
		}
```

As you can see, in this case postcondition `Program.ShouldBeTrue` was checked right after `builder.Start` method call but not after `builder.Task` is complete. This is a reason why this method fail during runtime and not correct.

> :question: Either way, why not check both synchronously? By this I mean either avoid a continuation where possible, or use TaskContinuationOptions.ExecuteSynchronously.

This is not a big deal to add additional `ExecuteSynchronously` to the continuation, but actual fix for this bug would be much more complicated.

**Comment by SergeyTeplyakov** [comment]
Ok. I've found a workaround for this... Will post an update shortly.

**Comment by sharwell** [comment]
> Second, I've posted incorrect decompiled version of the code. 

:arrow_up: This is the cause of my confusion. :smile:

I have a few other concerns regarding the actual implementation. I assume the continuation should only execute if the method completes successfully? If that is the case, the implementation should probably match the style used for [`CoreTaskExtensions.Catch`](https://github.com/rackerlabs/dotnet-threading/blob/master/Rackspace.Threading/CoreTaskExtensions.cs#L103-L130), or we run the risk of not propagating states correctly (in particular, the current continuation will convert a `Canceled` status to a `Faulted` status, and it also changes the exception stack traces).

**Comment by SergeyTeplyakov** [comment]
@sharwell Please, keep in mind that all those transformations are implemented on the IL level! And using any external libraries there would be super tricky!

**Comment by sharwell** [comment]
Yes, I'm not suggesting we actually use this library. I'm saying the generated IL needs to do the same thing or the rewriter will be changing the observable behavior of asynchronous methods.

**Comment by SergeyTeplyakov** [comment]
Current implementation is naive, but semantically it worked exactly like you proposed.

In the generated `CheckPost` method there is an access to `task.Result` witch will fail if the task will fail. This is not optimal, but this is relatively simple solution.

It should be relatively easy to execute continuation synchronously, but I can't see too much benefits right in changing the actual `CheckPost` method implementation.

**Comment by sharwell** [comment]
@SergeyTeplyakov The observable differences are:

1. If `task` is faulted, then accessing `task.Result` will throw an `AggregateException`. That `AggregateException` will get wrapped in its own additional `AggregateException`, completely changing the behavior of exception handling for callers.
2. If `task` is cancelled, then accessing `task.Result` will throw a `TaskCanceledException`, which will get converted to a `Faulted` state, not a `Canceled` state. When using the rewriter, all propagated information about a canceled task will be lost, potentially resulting in the execution of incorrect continuations.

**Comment by SergeyTeplyakov** [comment]
Yep, you're right! Will file separate bug and will propose the solution.

**Comment by sharwell** [comment]
@SergeyTeplyakov why not use this bug, which already has the proposed solution? :stuck_out_tongue_winking_eye:

**Comment by SergeyTeplyakov** [comment]
@sharwell We can, of course, but this is not the same bug:) So if someone will search for the real bug it would be almost impossible to understand why the fix to this bug changes something that is not described in the bug description:)

For this bug I would propose something as well and will close as wan't fix or by design:)
